folio-token/src/main/java/org/folio/spring/service/SystemUserService.java

```diff
@@ -0,0 +1,141 @@
+package org.folio.spring.service;
+
+import static java.util.Objects.isNull;
+import static org.folio.spring.utils.TokenUtils.parseUserTokenFromCookies;
+import static org.springframework.http.HttpHeaders.SET_COOKIE;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import java.time.Instant;
+import java.util.Optional;
+import java.util.function.Supplier;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.folio.spring.client.AuthnClient;
+import org.folio.spring.client.AuthnClient.UserCredentials;
+import org.folio.spring.client.UsersClient;
+import org.folio.spring.config.properties.FolioEnvironment;
+import org.folio.spring.context.ExecutionContextBuilder;
+import org.folio.spring.model.SystemUser;
+import org.folio.spring.model.UserToken;
+import org.folio.spring.scope.FolioExecutionContextSetter;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+import org.springframework.util.CollectionUtils;
+
+@Log4j2
+@Service
+@RequiredArgsConstructor
+public class SystemUserService {
+
+  private final ExecutionContextBuilder contextBuilder;
+  private final SystemUserProperties systemUserProperties;
+  private final FolioEnvironment environment;
+  private final AuthnClient authnClient;
+  private final PrepareSystemUserService prepareUserService;
+
+  private Cache<String, SystemUser> systemUserCache;
+
+  /**
+   * Get authenticate system user.
+   * Get from cache if present (or loginWithExpiry otherwise). Call loginWithExpiry token endpoint in case
+   * access token expired. Call login endpoint in case refresh token expired
+   *
+   * @param tenantId The tenant name
+   * @return {@link org.folio.spring.model.SystemUser} with token value
+   */
+  public SystemUser getAuthedSystemUser(String tenantId) {
+    if (systemUserCache == null) {
+      return getSystemUser(tenantId);
+    }
+
+    var user = systemUserCache.get(tenantId, this::getSystemUser);
+    var userToken = user.token();
+    var now = Instant.now();
+    if (userToken.accessTokenExpiration().isAfter(now)) {
+      return user;
+    }
+
+    systemUserCache.invalidate(tenantId);
+    if (userToken.accessTokenExpiration().isAfter(now)) {
+      var newToken = login(user.username());
```

Comment:
This `login` method only calls the old login endpoint. This is incorrect. It needs to first try `authn/login-wth-expiry`, then if that fails, try `authn/login`. You need a method that does both of these things, like Adam's client.

Both of the public methods in this class must use both login endpoints just like the vertx code.

Reply:
The `login` method here doesn't resolve. This code doesn't compile. To compile your code you need to `cd  folio-token` and then `mvn clean install -DskipTests` there. Once it compiles, run your tests by removing `-DskipTests`.

Reply:
Done